### PR TITLE
Support MLX in WhisperAX 

### PR DESCRIPTION
**Pull request by ZachNagengast**
Add support for the WhisperAX example app, as well as various refactors and cleanup.

image

Note: this is using an unreleased version of MLX, pending merge of https://github.com/ml-explore/mlx-swift/pull/130

There are still some memory issues to address, will look more into this soon. 

CoreML:
Screenshot 2024-09-06 at 11 28 23 PM
MLX:
image

Important note, first time running this will likely throw an error about PrepareMetalShaders, which requires Trust & Enable on this popup when selecting the error.
image

### CODE REVIEW

File: .swiftpm/configuration/Package.resolved [hunk 1]
```diff
@@ -28,6 +32,8 @@
       }
     },
     {
+=======
+>>>>>>> main
```
Comment (jkrukowski) [inline .swiftpm/configuration/Package.resolved:36]: git conflict markers

File: Package.resolved [hunk 2]
```diff
@@ -3,10 +3,9 @@
     {
       "identity" : "mlx-swift",
       "kind" : "remoteSourceControl",
-      "location" : "https://github.com/ml-explore/mlx-swift",
+      "location" : "https://github.com/davidkoski/mlx-swift.git",
       "state" : {
-        "revision" : "597aaa5f465b4b9a17c8646b751053f84e37925b",
-        "version" : "0.16.0"
+        "revision" : "3314bc684f0ccab1793be54acddaea16c0501d3c"
```
Comment (jkrukowski) [inline Package.resolved:8]: curious, why this change?

File: Sources/WhisperKit/Core/Models.swift [hunk 3]
```diff
@@ -171,6 +137,74 @@ public struct ModelComputeOptions {
     }
 }
 
+public struct ModelInfo: Identifiable, Hashable {
+    public let id = UUID()
```
Comment (jkrukowski) [inline Sources/WhisperKit/Core/Models.swift:141]: curious, why this property `public let id = UUID()` is needed? can it be uniquely identified by name?

File: Sources/WhisperKit/MLX/MLXTextDecoder.swift [hunk 4]
```diff
@@ -117,24 +126,59 @@ public final class MLXTextDecoder: TextDecoding {
...
+
+        let keyCache = try? MLX.stacked(keyCacheResult).asMLMultiArray()
+        let valueCache = try? MLX.stacked(valueCacheResult).asMLMultiArray()
+        let decodingCache = DecodingCache(
+            keyCache: keyCache,
+            valueCache: valueCache,
+            alignmentWeights: nil
+        )
+
+        let logits = try? result.logits?.asMLMultiArray()
```
Comment (jkrukowski) [inline Sources/WhisperKit/MLX/MLXTextDecoder.swift:152]: seems like you've changed it to `try ?` in couple of places, is it intended?

**Review by jkrukowski — COMMENTED** [review]
LGTM, added some comments, let me know what you think

File: Package.resolved (hunk 2 above)
Comment (jkrukowski) [reply Package.resolved:8]: ok, now I can see why, nvm

File: .swiftpm/configuration/Package.resolved (hunk 1 above)
Comment (ZachNagengast) [reply .swiftpm/configuration/Package.resolved:36]: Ah good catch 👍

File: Sources/WhisperKit/Core/Models.swift (hunk 3 above)
Comment (ZachNagengast) [reply Sources/WhisperKit/Core/Models.swift:141]: Needed to make it hashable, although I was using this for the picker at one point and may be vestigial, will check.

File: Sources/WhisperKit/MLX/MLXTextDecoder.swift (hunk 4 above)
Comment (ZachNagengast) [reply Sources/WhisperKit/MLX/MLXTextDecoder.swift:152]: Good point, we want this to pass the throw up through the call stack. Reviewing build warnings as well.

**Comment by iandundas** [comment]
Hey all, just a heads-up, the example project gets this SPM error:

CleanShot 2024-09-18 at 10 34 34@2x

**Comment by davidkoski** [comment]
> Hey all, just a heads-up, the example project gets this SPM error:
> 
> CleanShot 2024-09-18 at 10 34 34@2x

that fork was merged and is now the `0.16.2` tag on https://github.com/ml-explore/mlx-swift

**Comment by ZachNagengast** [comment]
Awesome, thanks for the update @davidkoski!

**Comment by latenitefilms** [comment]
@ZachNagengast - This looks awesome! Apologies - rookie question... It looks like currently the [MLX repo](https://huggingface.co/argmaxinc/whisperkit-coreml/blob/main/openai_whisper-base/config.json) only has the base and tiny models. How hard is it for mere mortals to "build" some of the larger models for testing this out with larger models?

**Comment by latenitefilms** [comment]
Or... can you use existing models with MLX?

**Comment by ZachNagengast** [comment]
Sorry missed your original message! We will fill in the mlx repo with the remaining models as part of this release, we just made these copies for consistency with our swift package. Any MLX whisper model currently existing with the same naming scheme will work in theory 👍  @jkrukowski may be able to confirm or deny.

**Comment by latenitefilms** [comment]
Legend, thanks @ZachNagengast! So basically, we do need new models for MLX, we can use the existing WhisperKit models? They need to be optimised or something?

**Comment by ZachNagengast** [comment]
Yep the existing WhisperKit models are optimized for CoreML, the ones in[ this repo](https://huggingface.co/argmaxinc/whisperkit-mlx/tree/main) we will fill out with the equivalent weights that are compatible with this MLX PR

**Comment by latenitefilms** [comment]
Sorry for all the rookie questions, but when you say "optimised for CoreML" - does this mean they **ONLY** work on CoreML, or can you use these CoreML models in MLX and they're just not as fast/accurate? 

Apologies - this whole Whisper world is very new to me, so I very much appreciate all your wisdom and support!

**Comment by maxlund** [comment]
> Yep the existing WhisperKit models are optimized for CoreML, the ones in[ this repo](https://huggingface.co/argmaxinc/whisperkit-mlx/tree/main) we will fill out with the equivalent weights that are compatible with this MLX PR

Is there any model conversion script we can run, or any other source we can use, in order to create/obtain more MLX compatible model versions?

**Comment by ZachNagengast** [comment]
@latenitefilms Yes the .mlmodelc models only work with CoreML at the moment. 
@maxlund There is a script made by @jkrukowski to do the conversion here https://github.com/argmaxinc/WhisperKit/pull/169, we'll integrate this into https://github.com/argmaxinc/whisperkittools in the future.

**Comment by latenitefilms** [comment]
Legend, thanks so much @ZachNagengast! Do you have a rough/ballpark ETA of when you're hoping to finish and merge in MLX support? No rush or pressure - just wondering if it's worth trying to convert our own models or not.

Let me know if there's anything I can do to help with MLX testing/release! Would love to see this in action ASAP!

Thanks for EVERYTHING you do! Appreciate it!

**Comment by ZachNagengast** [comment]
There are just a few optimizations to fix up to make it ready for release, specifically memory usage. Current issues are:
- MLX does not require a prewarm stage, so it should skip that. Currently its loading the model twice without freeing up the memory. Can also be solved by setting a cache limit or clearing the cache after load
- KV cache should use [this](https://github.com/ml-explore/mlx-swift-examples/blob/main/Libraries/LLM/KVCache.swift) instead of mlmultiarrays
- Sampling can be compiled for some easy speedups
- Attention should use [SDPA](https://swiftpackageindex.com/ml-explore/mlx-swift/main/documentation/mlxfast/scaleddotproductattention(queries:keys:values:scale:mask:stream:)) instead of [current logic](https://github.com/argmaxinc/WhisperKit/blob/f7a6c09864eed4427489bbdef3394ae3bde7c0c9/Sources/WhisperKit/MLX/Attention.swift#L59-L65)

These are paraphrased from @davidkoski and @awni

Will be revisiting this after the upcoming release but feel free to test with this current branch if you see any other potential speedup besides these, all the interfaces should be the same in its final form, just faster and more memory efficient with these changes.

**Comment by latenitefilms** [comment]
Amazing! Thanks so much! Will test out and let you know if I break anything.

**Comment by anishjain123** [comment]
Any update on this? Is it usable? @latenitefilms @ZachNagengast ? Can someone please point me in the direction of how i can get this set up?

**Comment by ZachNagengast** [comment]
Hi @anishjain123 these are still pending issues https://github.com/argmaxinc/WhisperKit/pull/200#issuecomment-2395207140, but this branch is technically usable. We'd like to resolve the perf and memory issues before merging, which is still a high priority for us! Working on a refactor right now to allow various different model input and output types, including MLXArray, which should help with the issues converting between MLMultiArray and MLXArray.

**Comment by ZachNagengast** [comment]
In order to consolidate a bit, I'll merge this PR into the parent branch so that there is one PR to manage the final changes in. Will link the relevant discussions there. If anyone has made progress on any of the topics discussed here please post in #124, there is some updating to do with the latest from main as well as mlx swift.

**Comment by maxlund** [comment]
@anishjain123 there is also this alternative, as you probably know already: [mlx-whisper](https://github.com/ml-explore/mlx-examples/tree/main/whisper)

@ZachNagengast what would be the differences between running Whisper via MLX using your implementation vs the mlx-whisper version (see link above)?
